python/pyproject.toml

```
@@ -43,25 +43,24 @@ include = [{ path = "dev", format = "sdist" }]
 
 [tool.poetry.dependencies]
 python = "^3.8"
-mmhash3 = "3.0.1"
-requests = ">=2.28.1,<=2.28.2"
-click = "8.1.3"
-rich = ">=13.0.0,<=13.3.4"
-pyyaml = ">=5.4.0,<=6.0.0"
-pydantic = "1.10.7"
-fsspec = ">=2022.8.2,<=2023.3.0"
-pyparsing = "3.0.9"
-zstandard = "0.20.0"
-pyarrow = { version = ">=8.0.0,<=11.0.0", optional = true }
-pandas = { version = ">=1.4.4,<=1.5.3", optional = true }
-duckdb = { version = ">=0.6.0,<=0.7.1", optional = true }
-ray = { version = ">=2.0.0,<=2.3.1", optional = true }
-python-snappy = { version = "0.6.1", optional = true }
-thrift = { version = "0.16.0", optional = true }
-boto3 = { version = "1.24.59", optional = true }
-# The versions of the fsspec implementations should run in sync with fsspec above
-s3fs = { version = ">=2022.8.2,<=2023.3.0", optional = true }
-adlfs = { version = ">=2022.8.2,<=2023.3.0", optional = true }
+mmhash3 = "^3"
+requests = "^2.20"
+click = ">=7.1.1,<9.0.0"
+rich = ">=10.11.0,<14.0.0"
+pyyaml = "^5.4" # CVE-2020-14343 was fixed in 5.4.
+pydantic = "^1.9"
+fsspec = ">=2021.09.0,<2024.1.0" # `lexists()` was implemented in 2021.09.0. Upper bound set arbitrarily, to be reassessed in early 2024.
+pyparsing = "^3.0.7" # The `min` keyword argument for `delimited_list()` was added in 3.0.7.
+zstandard = ">=0.13.0,<1.0.0"
+pyarrow = { version = ">=7.0.0,<12.0.0", optional = true }
+pandas = { version = ">=1.0.0,<3.0.0", optional = true }
+duckdb = ">=0.5.0,<1.0.0"
+ray = { version = "^2", optional = true }
+python-snappy = { version = "^0.6", optional = true }
+thrift = { version = ">=0.13.0,<1.0.0", optional = true }
+boto3 = { version = ">=1.17.106", optional = true }
+s3fs = { version = ">=2021.08.0,<2024.1.0", optional = true } # Upper bound set arbitrarily, to be reassessed in early 2024.
```

I think this should be in line with fsspec:
```suggestion
s3fs = { version = ">=2021.09.0,<2024.1.0", optional = true } # Upper bound set arbitrarily, to be reassessed in early 2024.
```

---

I previously added a comment as part of the justifications:

> `# TODO: Remove? Definitely doesn't seem to hold true to `adlfs`, and `s3fs` enforces it itself.`

To expand upon that, if you install `s3fs`, it will automatically bound the version of `fsspec` (in newer versions). On the other hand, `adlfs` does not release in lockstep with `fsspec`; e.g. the latest version there is `2023.1.0`, and its `fsspec` requirements is just `fsspec>=2021.10.1`.

---

I think removing would be better then.

---

Oh, sorry, rereading my comment I see I wasn't clear. The "Remove?" was regarding the comment saying the versions should be in lockstep; all of the dependencies still need to be kept (`fsspec` as a required dependency for the library with appropriate bounds, and the two extras for different filesystems).